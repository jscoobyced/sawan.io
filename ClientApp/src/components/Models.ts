import { match } from "react-router-dom";

export interface SelectOption {
    value: number;
    text: string;
}

export interface IdParam {
    id: string;
}
