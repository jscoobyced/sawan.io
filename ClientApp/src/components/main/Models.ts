import { NavigationMenuContent } from "../../services/Models";

export interface HeaderProps {
    navigationMenuContent: NavigationMenuContent;
    signIn: (isSignedIn: boolean) => void;
    isSignedIn: boolean;
}
