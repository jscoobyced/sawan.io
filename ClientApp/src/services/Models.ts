export enum Language {
    English = 0,
    French = 1
}

export interface Link {
    text: string;
    url: string;
    target?: string;
    title?: string;
}

export interface MainContent {
    navigationMenuContent: NavigationMenuContent;
    menuContent: MenuContent;
    footerContent: FooterContent;
}

export interface NavigationMenuContent {
    home: string;
    websiteName: string;
    applications: string;
    cryptoCurrency: string;
    healthMonitor: string;
    about: string;
    information: string;
    resume: string;
}

export interface BlogElement {
    id: number;
    blogDate: Date;
    article: string;
    articleTitle: string;
}

export interface BlogPage {
    articles: BlogElement[];
}

export interface MenuContent {
    links: Link[];
}

export interface FooterContent {
    copyright: string;
    credits: string;
    year: string;
}
