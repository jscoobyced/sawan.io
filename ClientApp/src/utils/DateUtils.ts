export class DateUtil {
    public static toHumanHours(date: Date): string {
        return `${this.toDigits(date.getHours(), 2)}:${this.toDigits(date.getMinutes(), 2)}`;
    }

    public static dateToString(date: Date): string {
        const lDate = new Date(date);
        const year = lDate.getFullYear();
        const month = DateUtil.toDigits(lDate.getMonth() + 1, 2);
        const day = DateUtil.toDigits(lDate.getDate(), 2);
        return `${year}-${month}-${day}`;
    }

    public static dateToYyyyMmString(date: Date): string {
        const lDate = new Date(date);
        const year = lDate.getFullYear();
        const month = DateUtil.toDigits(lDate.getMonth() + 1, 2);
        return `${year}${month}`;
    }

    public static defaultDate(now = false): Date {
        if (now) {
            const nowDate = new Date();
            const month = DateUtil.toDigits(nowDate.getMonth() + 1, 2);
            const day = DateUtil.toDigits(nowDate.getDate(), 2);
            return new Date(`${nowDate.getFullYear()}-${month}-${day}T00:00:00Z`);
        }

        return new Date('2019-01-01T00:00:00Z');
    }

    private static toDigits(value: number, length: number): string {
        let result: string = value.toString();
        while (result.length < length) {
            result = "0" + result;
        }

        return result;
    }
}
