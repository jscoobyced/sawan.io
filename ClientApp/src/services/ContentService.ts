import { ApiUtils } from '../utils/ApiUtils';
import { IContentService } from './IContentService';
import { BlogElement, BlogPage, Language, MainContent } from './Models';

export class ContentService implements IContentService {

    public async getMainContent(language: Language): Promise<MainContent> {
        return this.fetchMainContent(language);
    }

    public getDefaultMainContent(): MainContent {
        return this.defaultMainContent();
    }

    public async getBlogPage(yearMonth: string, maxResult: number): Promise<BlogPage> {
        return this.updateBlogPage(yearMonth, maxResult);
    }

    public async getBlogElement(id: number): Promise<BlogElement> {
        return ApiUtils.fetchData<BlogElement>(`/api/Data/blog/${id}`);
    }

    public getDefaultBlogPage(): BlogPage {
        return {
            articles: []
        };
    }

    public async saveBlogElement(blogElement: BlogElement): Promise<boolean> {
        return this.doSaveBlogElement(blogElement);
    }

    protected async doSaveBlogElement(blogElement: BlogElement): Promise<boolean> {
        return ApiUtils.postData('/api/Data/blog/post/', blogElement);
    }

    protected async fetchMainContent(language: Language): Promise<MainContent> {
        return ApiUtils.fetchData<MainContent>(`/api/Data/main/${language}`);
    }

    protected defaultMainContent(): MainContent {
        return {
            menuContent: {
                title: '',
                links: []
            },
            navigationMenuContent: {
                about: '',
                applications: '',
                cryptoCurrency: '',
                healthMonitor: '',
                home: '',
                websiteName: '',
                information: '',
                resume: ''
            },
            footerContent: {
                copyright: '',
                credits: '',
                year: ''
            }
        };
    }

    protected async updateBlogPage(yearMonth: string, maxResult: number): Promise<BlogPage> {
        const empty = {
            articles: []
        };
        if (maxResult <= 0) {
            return empty;
        }
        const data = await ApiUtils.fetchData<BlogElement[]>(`/api/Data/blogpage/${yearMonth}/${maxResult}`);
        if (!data || data.length === 0) {
            return empty;
        }
        return {
            articles: data.slice(0, Math.min(maxResult, data.length))
        };
    }
}
