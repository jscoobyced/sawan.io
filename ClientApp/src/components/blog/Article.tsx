import React from 'react';
import { NavLink } from 'react-router-dom';
import { BlogElement } from '../../services/Models';
import { DateUtil } from '../../utils/DateUtils';
import { HtmlUtils } from '../../utils/HtmlUtils';

export interface ArticleProps {
    blogElement: BlogElement;
    ellipsis?: boolean;
    backLink?: boolean;
    editLink?: boolean;
}

export class Article extends React.Component<ArticleProps> {
    public constructor(props: ArticleProps) {
        super(props);
    }

    public render() {
        const article = this.props.blogElement;
        if (!article) {
            return null;
        }
        let content = HtmlUtils.getSafeContent(article.article);
        if (this.props.ellipsis) {
            content = HtmlUtils.getEllipsis(content, 1000);
        }
        const blogDate = DateUtil.dateToString(!article.blogDate ? DateUtil.defaultDate() : article.blogDate);
        let editLink = null;
        let readMoreLink = null;
        let backLink = null;
        if (this.props.ellipsis) {
            readMoreLink = <NavLink to={'/blog/view/' + article.id} className='link-article'>Read more...</NavLink>;
        }
        if (this.props.backLink) {
            backLink = <NavLink to='/' className='btn-article'>Back</NavLink>;
        }
        if (this.props.editLink) {
            editLink = <NavLink to={'/blog/edit/' + article.id}>Edit</NavLink>;
        }

        return <div>
            <article className='blog-article'>
                <header>
                    <h1>{this.props.blogElement.articleTitle}</h1>
                    <span>{blogDate}</span>
                </header>
                <div className='article-content'>
                    <span dangerouslySetInnerHTML={{ __html: content }}></span>
                    <div className='article-links'>
                        {editLink}
                        {readMoreLink}
                    </div>
                </div>
                {backLink}
            </article>
        </div>;
    }
}
