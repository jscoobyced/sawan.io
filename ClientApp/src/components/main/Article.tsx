import React from 'react';
import { NavLink } from 'react-router-dom';
import { BlogElement } from '../../services/Models';
import { DateUtil } from '../../utils/DateUtils';
import { HtmlUtils } from '../../utils/HtmlUtils';

export interface ArticleProps {
    blogElement: BlogElement;
    ellipsis?: boolean;
    backLink?: boolean;
}

export class Article extends React.Component<ArticleProps> {
    public constructor(props: ArticleProps) {
        super(props);
    }

    public render() {
        const article = this.props.blogElement;
        if (!article) {
            return null;
        }
        let content = HtmlUtils.getSafeContent(article.article);
        if (this.props.ellipsis) {
            content = HtmlUtils.getEllipsis(content, 1000);
        }
        const blogDate = DateUtil.dateToString(article.blogDate);
        return <div>
            <article>
                <h1>{this.props.blogElement.articleTitle} - {blogDate}</h1>
                <div>
                    <span dangerouslySetInnerHTML={{ __html: content }}></span>
                    {
                        this.props.ellipsis
                            ? <NavLink to={'/blog/' + article.id} className='link-article'>Read more...</NavLink>
                            : null
                    }
                </div>
            </article>
            {
                this.props.backLink
                    ? <NavLink to='/' className='link-article'>Back</NavLink>
                    : null

            }
        </div>;
    }
}