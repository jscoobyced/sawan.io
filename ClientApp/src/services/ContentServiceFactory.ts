import { ContentService } from './ContentService';
import { IContentService } from './IContentService';
import { MockContentService } from './MockContentService';

export class ContentServiceFactory {

    public static GetContentService(): IContentService {
        const mode = process.env.mode as string;
        if (mode === 'production') {
            return new ContentService();
        }

        return new MockContentService();
    }
}
