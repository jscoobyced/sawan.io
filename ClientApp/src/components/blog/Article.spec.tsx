import { shallow } from 'enzyme';
import React from 'react';
import { BlogElement } from '../../services/Models';
import { DateUtil } from '../../utils/DateUtils';
import { Article, ArticleProps } from './Article';

const blogElement: BlogElement = {
  article: 'bla bla bla',
  articleTitle: 'Title',
  blogDate: DateUtil.defaultDate(),
  updateDate: DateUtil.defaultDate(),
  id: 1,
};
const data: ArticleProps[] = [
  { blogElement, ellipsis: true, backLink: true },
  { blogElement, ellipsis: false, backLink: true },
  { blogElement, ellipsis: true, backLink: false },
  { blogElement, ellipsis: false, backLink: false },
  {
    blogElement, ellipsis: false, backLink: false, editLink: true,
  },
];


describe('Article', () => {
  it('Article component with empty article.', async () => {
    const article = shallow(<Article blogElement={null as unknown as BlogElement} />);
    expect(article.find('article')).toHaveLength(1);
  });

  it('Article component with article.', async () => {
    data.forEach(articleProps => {
      const article = shallow((
        <Article
          blogElement={articleProps.blogElement}
          ellipsis={articleProps.ellipsis}
          backLink={articleProps.backLink}
          editLink={articleProps.editLink}
        />));
      expect(article.find('article')).toHaveLength(1);
    });
  });
});
