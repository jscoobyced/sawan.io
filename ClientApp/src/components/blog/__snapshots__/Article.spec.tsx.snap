// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article component with article. 1`] = `
<div>
  <article
    className="blog-article"
  >
    <header>
      <h1>
        Title
      </h1>
      <span>
        2019-01-01
      </span>
    </header>
    <div
      className="article-content"
    >
      <span
        dangerouslySetInnerHTML={
          Object {
            "__html": "bla bla bla",
          }
        }
      />
      <div
        className="article-links"
      >
        <NavLink
          className="link-article"
          to="/blog/view/1"
        >
          Read more...
        </NavLink>
      </div>
    </div>
    <NavLink
      className="btn-article"
      to="/"
    >
      Back
    </NavLink>
  </article>
</div>
`;

exports[`Article component with article. 2`] = `
<div>
  <article
    className="blog-article"
  >
    <header>
      <h1>
        Title
      </h1>
      <span>
        2019-01-01
      </span>
    </header>
    <div
      className="article-content"
    >
      <span
        dangerouslySetInnerHTML={
          Object {
            "__html": "bla bla bla",
          }
        }
      />
      <div
        className="article-links"
      />
    </div>
    <NavLink
      className="btn-article"
      to="/"
    >
      Back
    </NavLink>
  </article>
</div>
`;

exports[`Article component with article. 3`] = `
<div>
  <article
    className="blog-article"
  >
    <header>
      <h1>
        Title
      </h1>
      <span>
        2019-01-01
      </span>
    </header>
    <div
      className="article-content"
    >
      <span
        dangerouslySetInnerHTML={
          Object {
            "__html": "bla bla bla",
          }
        }
      />
      <div
        className="article-links"
      >
        <NavLink
          className="link-article"
          to="/blog/view/1"
        >
          Read more...
        </NavLink>
      </div>
    </div>
  </article>
</div>
`;

exports[`Article component with article. 4`] = `
<div>
  <article
    className="blog-article"
  >
    <header>
      <h1>
        Title
      </h1>
      <span>
        2019-01-01
      </span>
    </header>
    <div
      className="article-content"
    >
      <span
        dangerouslySetInnerHTML={
          Object {
            "__html": "bla bla bla",
          }
        }
      />
      <div
        className="article-links"
      />
    </div>
  </article>
</div>
`;

exports[`Article component with empty article. 1`] = `""`;
